Firestore/core/src/remote/grpc_connection.cc

```diff
@@ -120,6 +121,22 @@ HostConfigMap& Config() {
   return config_by_host_;
 }
 
+#if __APPLE__
+// Disable CFStream-based transport on Apple platforms due to b/133182964, where
+// CFStream will occasionally fail to raise a has-bytes-available events,
+// causing Firestore to appear to hang.
+//
+// Use a constructor of a globally scoped object to set the `grpc_cfstream`
```

Comment:
This is ingenious, though I'd like to double-check if it's necessary. Does the gRPC team recommend doing it this way? Is there perhaps some well-defined point better suited for `setenv` calls? Note that this is mostly me playing the devil's advocate -- I'm pretty sure this is a good implementation.

Reply:
gRPC team didn't have any more advice than "set `grpc_cfstream=0`". Here's their official documentation: https://github.com/grpc/grpc/blob/master/src/objective-c/README-CFSTREAM.md

In general, the best way to handle calls to `setenv` is to put them at the top of your `main` function. This avoids data races that are otherwise inherent in multiple threads setting and getting environment variables concurrently. As a library author we don't really control `main` so it's kind of a crapshoot as to whether or not we'll make the call before anyone else uses gRPC.

Indeed, my initial implementation was to put this in the `GrpcConnection` constructor:

```c++
#if __APPLE__
  static std::once_flag grpc_cfstream_disabled;

  std::call_once(grpc_cfstream_disabled, [] {
    setenv("grpc_cfstream", "0", 1);
  });
#endif  // __APPLE__
```

However, this didn't work. In gRPC-C++, the call to `grpc_init` happens implicitly and it seems to happen for essentially any API call. Other calls must have happened earlier and prevented the variable from being read.

I looked for other places insert this to make this earlier (for example, making a static `Datastore::InitGrpc` and calling it from the `api::Firestore` constructor (or similar) but this causes the problem that every different way we construct Firestore or FirestoreClient instances has to be adjusted or it will be subject to flakes if we forget. I didn't want to take the chance that we might forget to make this call so this approach seemed better.

Reply:
Thanks for elaborating on the rationale. I think it's a great solution.